src/Servers/Kestrel/Core/src/Internal/Infrastructure/StringUtilities.cs

```diff
@@ -109,6 +113,222 @@ public static unsafe bool TryGetAsciiString(byte* input, char* output, int count
+                        {
+                            goto NotEqual;
+                        }
+                    }
+
+                    // End of input reached, there are no inequalities via widening; so the input bytes are both ascii
+                    // and a match to the string if it was converted via Encoding.ASCII.GetString(...)
+                    return true;
+                }
+
+                // Create a comparision vector for all bits being equal
+                var AllTrue = new Vector<ushort>(ushort.MaxValue);
+                // do/while as entry condition already checked, remaining length must be Vector<byte>.Count or larger.
+                do
+                {
+                    // Read a Vector length from the input as bytes
+                    var vector = Unsafe.ReadUnaligned<Vector<byte>>(ref Unsafe.Add(ref bytes, offset));
+                    // Widen the bytes directly to chars (ushort) as if they were ascii.
+                    // As widening doubles the size we get two vectors back.
+                    Vector.Widen(vector, out var vector0, out var vector1);
+                    // Read two char vectors from the string to perform the match.
+                    var compare0 = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref str, offset)));
+                    var compare1 = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref str, offset + Vector<ushort>.Count)));
+
+                    // If the string is not ascii, then the widened bytes cannot match
+                    // as each widened byte element as chars will be in the range 0-255
+                    // so cannot match any higher unicode values.
+
+                    // Compare to our all bits true comparision vector
+                    if (!AllTrue.Equals(
+                        // BitwiseAnd the two equals together
+                        Vector.BitwiseAnd(
+                            // Check equality for the two widened vectors
+                            Vector.Equals(compare0, vector0),
+                            Vector.Equals(compare1, vector1))))
+                    {
+                        goto NotEqual;
+                    }
+
+                    offset += Vector<byte>.Count;
+                } while ((byte*)(offset + Vector<byte>.Count) <= (byte*)count);
+
+                // Vector path done, loop back to do non-Vector
+                // If is a exact multiple of vector size, bail now
+            } while ((byte*)offset < (byte*)count);
+
+            // If we get here (input is exactly a multiple of Vector length) then there are no inequalities via widening;
+            // so the input bytes are both ascii and a match to the string if it was converted via Encoding.ASCII.GetString(...)
+            return true;
+        NotEqual:
+            return false;
+        }
+
+        private unsafe static bool IsValidHeaderString(string value)
```

Comment:
This method is only called within `Debug.Assert` and doesn't need to be optimized. The entire method can be implemented as below:

```cs
try {
  if (value is null) return false;
  new Utf8Encoding(throw: true).GetByteCount(value);
  return true;
} catch (...) {
  return false;
}
```

Reply:
Done